### PR TITLE
Explain the firewall between query errors and data mismatches

### PR DESCRIPTION
**Pull request by cabo**
(as per AD review by Murray Kucherawy)

### CODE REVIEW

File: draft-ietf-jsonpath-base.md [hunk 1]
```diff
@@ -541,6 +541,16 @@ The well-formedness and the validity of JSONPath queries are independent of
 the JSON value the query is applied to. No further errors relating to the
 well-formedness and the validity of a JSONPath query can be
 raised during application of the query to a value.
+This clearly separates well-formedness/validity errors in the query
+from mismatches that may actually stem from flaws in the data (see
+also {{data-attack}}).
+
+Often, mismatches between the structure expected by a valid query
```
Comment (timbray) [inline draft-ietf-jsonpath-base.md:548]: Suggest: It is possible that mismatches … in the data lead to empty query results.

(Don't think "often" is called for, any path to replace it is OK by me)

File: draft-ietf-jsonpath-base.md [hunk 2]
```diff
@@ -541,6 +541,16 @@ The well-formedness and the validity of JSONPath queries are independent of
...
+and the structure found in the data will lead to empty query results,
+which may be unexpected and indicate bugs in either.
+JSONPath implementations might therefore want to provide diagnostics
```
Comment (timbray) [inline draft-ietf-jsonpath-base.md:551]: suggest MAY?

File: draft-ietf-jsonpath-base.md [hunk 3]
```diff
@@ -2212,6 +2225,20 @@ The resulting class of attacks, *injections* (e.g., SQL injections),
 is consistently found among the top causes of application security
 vulnerabilities and requires particular attention.
 
+## Attack Vectors on the Data Examined by JSONPath Queries {#data-attack}
```
Comment (timbray) [inline draft-ietf-jsonpath-base.md:2228]: Really?  I'm having trouble believing this.  I think you should lose this and stick with the good wording up at the top of the review.  This is really extraneous to Murray's comment.

File: draft-ietf-jsonpath-base.md [hunk 4]
```diff
@@ -2167,7 +2179,8 @@ and the entries in the column
 Security considerations for JSONPath can stem from
 
 * attack vectors on JSONPath implementations,
-* attack vectors on how JSONPath queries are formed, and
+* attack vectors on how JSONPath queries are formed
```
Comment (timbray) [inline draft-ietf-jsonpath-base.md:2182]: I don't think this change adds any value.

**Review by timbray — CHANGES_REQUESTED** [review]

File: draft-ietf-jsonpath-base.md (hunk 2 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md:551]: This is not an interoperability requirement.  It doesn't affect the other side.  We use MAY to say party A MAY do something, so party B ~~needs to~~ MUST be prepared for that behavior.  That's why I believe this is a "might" (non-2119 may).

File: draft-ietf-jsonpath-base.md (hunk 1 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: I went for a simple "can".

**Review by timbray — APPROVED** [review]

File: draft-ietf-jsonpath-base.md (hunk 3 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md:2228]: Well, it was the main reason *I* care for the clear separation between query and data errors.
So I don't think any explanation of this is complete without this.
But I agree we don't *have* to explain all that.

Comment (timbray) [reply draft-ietf-jsonpath-base.md:2228]: Murray asked for a little bit of explanation for a basic fact about JSONPath that frequently surprises people.  I think adding a new security consideration is overkill, and also it's a significant change that I think the WG should get to comment on.  So I still think we should just skip this.

**Review by timbray — APPROVED** [review]

File: draft-ietf-jsonpath-base.md (hunk 3 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: I split out the seccons part to #493, so #492 should now be clean of security

**Review by glyn — APPROVED** [review]
Happy to include this or not. There are lots of unspecified behaviours implementations could provide. I think this probably adds some value to the spec, though.

**Pull request merged**